library/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/async/RecordedDataQueueHandler.kt

```diff
@@ -0,0 +1,177 @@
+) {
+
+    // region internal
+    internal var recordedDataQueue = ConcurrentLinkedQueue<RecordedDataQueueItem>()
+
+    @MainThread
+    internal fun addTouchEventItem(
+        pointerInteractions: List<MobileSegment.MobileRecord>
+    ): TouchEventRecordedDataQueueItem? {
+        val rumContextData = rumContextDataHandler.createRumContextData()
+            ?: return null
+
+        val item = TouchEventRecordedDataQueueItem(
+            timestamp = rumContextData.timestamp,
+            prevRumContext = rumContextData.prevRumContext,
+            newRumContext = rumContextData.newRumContext
+        )
+
+        item.touchData = pointerInteractions
+
+        insertIntoRecordedDataQueue(item)
+
+        return item
+    }
+
+    @MainThread
+    internal fun addSnapshotItem(
+        systemInformation: SystemInformation
+    ): SnapshotRecordedDataQueueItem? {
+        val rumContextData = rumContextDataHandler.createRumContextData()
+            ?: return null
+
+        val item = SnapshotRecordedDataQueueItem(
+            timestamp = rumContextData.timestamp,
+            prevRumContext = rumContextData.prevRumContext,
+            newRumContext = rumContextData.newRumContext,
+            systemInformation = systemInformation
+        )
+
+        item.systemInformation = systemInformation
+
+        insertIntoRecordedDataQueue(item)
+
+        return item
+    }
+
+    /**
+     * Goes through the queue one item at a time for as long as there are items in the queue.
+     * If an item is ready to be consumed, it is processed.
+     * If an invalid item is encountered, it is removed (invalid items are possible
+     * for example if a snapshot failed to traverse the tree).
+     * If neither of the previous conditions occurs, the loop breaks.
+     */
+    @MainThread
+    internal fun tryToConsumeItems() {
+        // no need to create a thread if the queue is empty
+        if (recordedDataQueue.isEmpty()) {
+            return
+        }
+
+        // currentTime needs to be obtained on the uithread
+        val currentTime = timeProvider.getDeviceTimestamp()
+
+        @Suppress("SwallowedException", "TooGenericExceptionCaught")
+        try {
+            executorService.execute {
+                triggerProcessingLoop(currentTime)
+            }
+        } catch (e: RejectedExecutionException) {
+            // TODO: REPLAY-1364 Add logs here once the sdkLogger is added
+        } catch (e: NullPointerException) {
+            // in theory will never happen but we'll log it
+            // TODO: REPLAY-1364 Add logs here once the sdkLogger is added
+        }
+    }
+
+    @WorkerThread
+    private fun triggerProcessingLoop(currentTime: Long) {
+        while (recordedDataQueue.isNotEmpty()) {
+            val nextItem = recordedDataQueue.peek()
+
+            if (nextItem != null) {
+                if (shouldRemoveItem(nextItem, currentTime)) {
+                    recordedDataQueue.poll()
```

Comment:
that's the reason I suggested to use `remove(item)` instead of `poll` and on top of that I also agree with the synchronisation.

Reply:
if method is synchronised - yes. But instead we can put a synchronised lock only on the parts of the method which modify data and do a double check (inside and outside synchronised block), don't know which one will be better. I guess synchronising the whole method is not needed, since we are loosing the efficiency of the parallelism for the parts where queue is not modified.

Reply:
actually we can be sure that we can never encounter this sort of issue, because the threadpool has a size/maxsize of 1 so there will never be more than one thread executing at any one time. In fact, we don't even need to synchronise the method because this makes it implicitly synchronised

Reply:
> the threadpool has a size/maxsize of 1 so there will never be more than one thread executing at any one time

Technically, the maximum pool size can be increased with a simple `ThreadPoolExecutor.setMaximumPoolSize()`. Also from this class's point of view, this method is triggered by an `ExecutorService`, so we have no guarantee of what the actual implementation is going to be.

Reply:
That's a good point. I've discussed this with Marius, and what I'm going to do is to create two constructors for the class - a regular constructor where we internally create the executorService, and another constructor visible only for testing where we pass in the executor. In this way we can be fairly sure that we encapsulate the executor and that we're only ever using a single thread, and avoid the added overhead of synchronising the method

Reply:
I guess the concern here is that if at any point someone does the change to the max number of threads in the pool by changing the constant or by calling the `setMaximumPoolSize` things will start to be buggy and will be tricky to catch the root cause.

This algorithm is working fine if there is only a single thread accessing the queue at the time. I'm okay with that if we put some checks/restrictions to make sure that will always be the case. Otherwise we need to change the way we are working with the queue.

For the moment, to make sure that this piece of code will be fine in any scenario, you can just make top-level method `synchronised`, but for sure if we at some point allow multiple threads to work with the queue, better locking approach should be developed.